keps/sig-api-machinery/2019-02-13-custom-resource-definitions.md

```diff
@@ -0,0 +1,212 @@
+In this usage, the CustomResource is an abstraction that "compiles" down to lower level Kubernetes
+primitives. In this pattern there is a "compiler" process that is running in the cluster, watching
+for new resources to be created and "compiling" them into Kubernetes primitives, and likewise
+looking for modifications and deletes and reacting accordingly, but beyond this compiler, there is
+no control loop running to dynamically manage the objects created from the CustomResourceDefinition.
+It is assumed that lower level Kubernetes objects (e.g. `ReplicaSet`) are used to achieve self-healing
+behaviors if necessary.
+
+#### Operator Patterns
+
+The operator pattern is well documented in a variety of places. It extends from the compiler pattern
+to include online maintainence and repair of the resources created by the CustomResource. Operators are processes that watch their respective custom resources and take online actions to ensure the health of the custom resource. An example
+of such repair might be a `Database` custom resource. It follows the compiler pattern by producing a
+`StatefulSet` to deploy the database, but it extends that pattern to become an operator, because it
+also performs periodic backups of the database to cloud storage, and it can recover from those
+backups if all Pods in the StatefulSet fail. 
+
+### Open Questions
+
+#### Dependency management
+
+If truly core APIs are developed as custom resources, then there can be dependency issues when
+bootstrapping a Kubernetes cluster. If core components (e.g. the controller manager) needs to manage
+a custom resource, how do you ensure that the resource gets created before the controller-manager
+starts, when the controller-manager itself is in the lifecycle for custom resources. We need to untangle
+these circular dependencies. And ensure that people who write such controllers, write them in such a
+way that they can temporarily handle the resources they manage not existing.
+
+Another significant concern in dependency management is dependencies between custom resource definitions.
+As the ecosystem of plugins becomes richer, the chances that one plugin will depend on another increases.
+Currently there is no good way to express that in order for CustomResource `A` to work correctly, 
+CustomResource `B` needs to be installed as well.
+
+#### Version upgrades
+
+Another open question for CustomResources is version management. Currently there are no built-in practices
+for moving from one version of a CustomResource to another. To do such a migration, the tool itself would
+have to perform the schema migration by reading all of the data objects in the old version and rewriting
+them in the new version. To make this successful , at the very least we will need to provide the ability to
+mark a CustomResource as "read-only" to ensure that there isn't skew between the two representations.
```

Comment:
Should conversion webhooks be mentioned here, or is that out of scope for this doc?